pkg/controller/controller.go

```diff
@@ -765,6 +741,220 @@ func (c *M3DBController) handleClusterUpdate(cluster *myspec.M3DBCluster) error
+	cluster *myspec.M3DBCluster,
+	actual *appsv1.StatefulSet,
+	expected *appsv1.StatefulSet,
+	update bool,
+) (bool, error) {
+	if !update {
+		return true, nil
+	}
+
+	_, err := c.kubeClient.AppsV1().StatefulSets(cluster.Namespace).Update(expected)
+	if err != nil {
+		c.logger.Error(err.Error())
+		return false, err
+	}
+
+	c.logger.Info("updated statefulset",
+		zap.String("name", expected.Name),
+		zap.Int32("actual_readyReplicas", actual.Status.ReadyReplicas),
+		zap.Int32("actual_updatedReplicas", actual.Status.UpdatedReplicas),
+		zap.String("actual_currentRevision", actual.Status.CurrentRevision),
+		zap.String("actual_updateRevision", actual.Status.UpdateRevision),
+		zap.Int32("expected_readyReplicas", expected.Status.ReadyReplicas),
+		zap.Int32("expected_updatedReplicas", expected.Status.UpdatedReplicas),
+		zap.String("expected_currentRevision", expected.Status.CurrentRevision),
+		zap.String("expected_updateRevision", expected.Status.UpdateRevision),
+		zap.Int64("generation", expected.Generation),
+		zap.Int64("observed", expected.Status.ObservedGeneration),
+	)
+
+	return false, err
+}
+
+func (c *M3DBController) updateWithOnDeleteStrategy(
+	cluster *myspec.M3DBCluster,
+	actual *appsv1.StatefulSet,
+	expected *appsv1.StatefulSet,
+	update bool,
+) (bool, error) {
+	logger := c.logger.With(
+		zap.String("namespace", cluster.Namespace), zap.String("name", actual.Name),
+	)
+
+	if update {
+		updated, err := c.kubeClient.AppsV1().StatefulSets(cluster.Namespace).Update(expected)
+		if err != nil {
+			logger.Error("error updating statefulset", zap.Error(err))
+			return false, err
+		}
+		actual = updated
+
+		c.logger.Info("updated statefulset",
+			zap.String("name", expected.Name),
+			zap.Int32("actual_readyReplicas", actual.Status.ReadyReplicas),
+			zap.Int32("actual_updatedReplicas", actual.Status.UpdatedReplicas),
+			zap.String("actual_currentRevision", actual.Status.CurrentRevision),
+			zap.String("actual_updateRevision", actual.Status.UpdateRevision),
+			zap.Int32("expected_readyReplicas", expected.Status.ReadyReplicas),
+			zap.Int32("expected_updatedReplicas", expected.Status.UpdatedReplicas),
+			zap.String("expected_currentRevision", expected.Status.CurrentRevision),
+			zap.String("expected_updateRevision", expected.Status.UpdateRevision),
+			zap.Int64("generation", expected.Generation),
+			zap.Int64("observed", expected.Status.ObservedGeneration),
+		)
+	} else if _, ok := actual.Annotations[annotations.ParallelUpdateInProgress]; !ok {
+		logger.Debug("no update and no rollout in progress so move to next statefulset")
+		return true, nil
+	}
+
+	numPods, err := c.getMaxPodsToUpdate(actual)
+	if err != nil {
+		logger.Error("error retrieving number of pods to update in parallel", zap.Error(err))
+		return false, err
+	}
+
+	if numPods == 0 {
+		err = errors.New("parallel update annotation set to 0. will not perform pod updates")
+		logger.Error(err.Error())
```

Comment:
is it really necessary to log the errors in this method? I would think something higher up would be logging the errors

Reply:
Yeah, this is generally the pattern I observed in the operator. Traced the call stack to where these errors are handled and nothing does.

Reply:
yea nvm, this is cause we don't propagate call stacks in the operator.

Reply:
Actually, nevermind. Will move this logging up to the loop that invokes the updates depending on the update strategy which should help a bit